### PR TITLE
Hw 1

### PR DESCRIPTION
**Pull request by onevoker**

### CODE REVIEW

File: bot/src/main/java/edu/java/bot/commands/ListCommand.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+        if (userLinks == null || userLinks.isEmpty()) {
+            text = NOT_LINKED_MESSAGE;
+            return new SendMessage(chatId, text.toString());
+        }
+        text = new StringBuilder(HANDLE_TEXT);
+        int i = 0;
+
+        for (URI link : userLinks) {
+            text.append("%s. ".formatted(++i)).append(link.toString()).append("\n");
+        }
```
Comment (iskanred) [inline bot/src/main/java/edu/java/bot/commands/ListCommand.java:46]: лучше использовать какую-нибудь либку типа Joiner, либо вынести эту логику в новую приватную функцию:
`
private String getLinksListText(Collection<URI> links) {
    ...
}
`

Comment (onevoker) [reply bot/src/main/java/edu/java/bot/commands/ListCommand.java]: Вынес логику в приватную функцию, после коммита у меня почему-то в классе LinkTest упал тест testCorrectUri, локально у меня все тесты проходят, не могу понять в чем проблема, я пока на этот тест поставил @Disabled

Comment (onevoker) [reply bot/src/main/java/edu/java/bot/commands/ListCommand.java]: Вынес логику в приватную функцию, после коммита у меня почему-то в классе LinkTest упал тест testCorrectUri, локально у меня все тесты проходят, не могу понять в чем проблема, я пока этот тест отключил

File: bot/src/main/java/edu/java/bot/links/Link.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+            if (statusCode != AVAILABLE_RESPONSE_CODE) {
+                return false;
+            }
+            return isGitHubOrStackOverflowLink(link);
+        } catch (IOException | InterruptedException | URISyntaxException | IllegalArgumentException e) {
+            return false;
+        }
+    }
+
+    private static boolean isGitHubOrStackOverflowLink(String link) {
```
Comment (iskanred) [inline bot/src/main/java/edu/java/bot/links/Link.java:34]: лучше назвать по типу `isValidReource`
при добавлении новых источников, названия функции тоже менять придется иначе)

File: bot/src/main/java/edu/java/bot/links/Link.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+import java.io.IOException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.http.HttpClient;
+import java.net.http.HttpRequest;
+import java.net.http.HttpResponse;
+import java.util.Objects;
+
+public class Link {
```
Comment (iskanred) [inline bot/src/main/java/edu/java/bot/links/Link.java:11]: Я бы подумал в сторону реализации ссылки, как модели, котора бы  в себе хранило строковое представление или `URI`. Проверки на валидность можно сделать прям в конструкторе с выбросом исключения. Либо отдельно сделать метод `validate`  или `checkValidity` или `isValid` или `Optional<Link> takeIfValid()`.

Если все-таки предпочтешь оставить утилитарный класс со статическими методами, то назвать надо соответственно `LinkUtils`, `LinkChecker` и т.п.

File: bot/src/main/java/edu/java/bot/links/UserLinks.java [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+    public void addUserLink(User user, String link) {
+        Long userId = user.id();
+        Set<URI> set = links.computeIfAbsent(userId, k -> new HashSet<>());
+        set.add(normalizeLink(link));
+    }
+
+    public void deleteUserLink(User user, String link) {
+        Set<URI> set = links.get(user.id());
+        if (!set.isEmpty()) {
```
Comment (iskanred) [inline bot/src/main/java/edu/java/bot/links/UserLinks.java:37]: проверку делать необязательно, если элемента в сете нет, `remove` просто вернет `false`

File: bot/src/main/java/edu/java/bot/links/UserLinks.java [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import org.springframework.stereotype.Component;
+
+@Component
+public class UserLinks {
```
Comment (iskanred) [inline bot/src/main/java/edu/java/bot/links/UserLinks.java:13]: naming: можно назвать `LinksRepository`

Подробнее про паттерн репозиторий:
https://habr.com/ru/articles/248505/

File: bot/src/main/java/edu/java/bot/links/UserLinks.java [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+    private final Map<Long, Set<URI>> links;
+
+    public UserLinks() {
+        this.links = new HashMap<>();
+    }
+
+    public boolean isRegistered(User user) {
```
Comment (iskanred) [inline bot/src/main/java/edu/java/bot/links/UserLinks.java:20]: лучше разнести фунциональность по пользователям и ссылкам в отдельные классы

```java
class UserRepository {
    // user-id: user
    private Map<Integer, User> users; // или Set<User> users и переопределение equals и hashcode в User
}

class LinkRepository {
    private Map<Ingteger, Link> links; // или опять же Set<Link>
}

```
При этом сам класс Link тогда должен будет содержать user-id в себе

Comment (iskanred) [reply bot/src/main/java/edu/java/bot/links/UserLinks.java:20]: Но это сложновато может быть, просто в будуших домашках придется разносить все равно, можешь пока так оставить)

File: bot/src/main/java/edu/java/bot/links/UserLinks.java [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+            return this.getUserLinks(user).contains(normalizeLink(link));
+        } catch (Exception e) {
+            return false;
+        }
+    }
+
+    private URI normalizeLink(String link) {
+        try {
+            return new URI(link.replaceAll("/+$", ""));
+        } catch (URISyntaxException e) {
```
Comment (iskanred) [inline bot/src/main/java/edu/java/bot/links/UserLinks.java:58]: так делать не стоит)
потом будет непонятно, в чем конкретно ошибка. Исключения должны явно указывать на то, что пошло не так.
Если не хочешь пробрасывать check excception, можешь создать свой класс, который будет наследоваться от `RuntimeException` и назвать как-то типа `InvalidLinkException`

File: bot/src/main/java/edu/java/bot/processor/UserMessageProcessorImpl.java [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+import java.util.List;
+
+public class UserMessageProcessorImpl implements UserMessageProcessor {
+    private final UserLinks links = new UserLinks();
+    private static final String UNKNOWN_COMMAND_TEXT =
+        "Мне не известна эта команда, для получения доступных команд воспользуйтесь командой /help";
+
+    @Override
+    public List<? extends Command> commands() {
+        return List.of(
```
Comment (iskanred) [inline bot/src/main/java/edu/java/bot/processor/UserMessageProcessorImpl.java:23]: Тут list каждый раз заново создается при кажом вызове? Можно ли его хранить, как поле для инстанса `UserMessageProcessorImpl`?

File: bot/src/main/resources/application.yml [hunk 9]
```diff
@@ -1,5 +1,5 @@
 app:
-  telegram-token: unset
+  telegram-token: ${tg-token}
```
Comment (iskanred) [inline bot/src/main/resources/application.yml:2]: ✅

File: bot/src/test/java/bot/commandsTest/HelpCommandTest.java [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.InjectMocks;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+import static org.assertj.core.api.AssertionsForClassTypes.assertThat;
+import static org.mockito.Mockito.doReturn;
+
+class HelpCommandTest {
+    @ExtendWith(MockitoExtension.class)
```
Comment (iskanred) [inline bot/src/test/java/bot/commandsTest/HelpCommandTest.java:17]: Не встречал такой записи) Это равносильно тому, чтобы над классом аннотацю поставить? Если да, то лучше над классом

File: bot/src/test/java/bot/commandsTest/HelpCommandTest.java [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+
+    @Test
+    void testCommand() {
+        String nameOfCommand = helpCommand.command();
+        String expected = "/help";
+        assertThat(nameOfCommand).isEqualTo(expected);
+    }
+
+    @Test
+    void testDescription() {
```
Comment (iskanred) [inline bot/src/test/java/bot/commandsTest/HelpCommandTest.java:43]: тест немного избыточный, он просто проверяет, что поле в классе дейтсвительно равно тому, что там написано

если описание будут менять, то и тест придется менять, и смысла в тесте становится немного

File: bot/src/test/java/bot/commandsTest/HelpCommandTest.java [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+    @ExtendWith(MockitoExtension.class)
+
+    @Mock
+    private Message message;
+    @Mock
+    private Chat chat;
+    @Mock
+    private Update update;
+
+    private void setUpTest() {
```
Comment (iskanred) [inline bot/src/test/java/bot/commandsTest/HelpCommandTest.java:26]: поскольку функция используется только в одном тесте, то лучше прям так и внести ее тело в этот тест

Когда смотришь на тесты, хочется понимать три вещи:
1) Какое окружение конкретно для этого теста было настроено (что замокировано, что хотим возвращать)
2) Что проверям? - Вызов самого метода
3) Как проверяем? - Сранвниваем expected с actual значениями

File: bot/src/test/java/bot/commandsTest/TrackCommandTest.java [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+    private Chat chat;
+    @Mock
+    private Update update;
+    private final UserLinks links = new UserLinks();
+    @InjectMocks
+    private Command trackCommand = new TrackCommand(links);
+    private static final User USER = new User(-10L);
+    private static final String GIT_HUB = "https://github.com/onevoker";
+
+    private void setUpTest(String returnedTextFromMessage) {
```
Comment (iskanred) [inline bot/src/test/java/bot/commandsTest/TrackCommandTest.java:34]: Тут `setUptTest` уже нужно больше, чем для одного теста. Но исходя из названии функции непонятно, что оно сетапит и почему не для всех тестов используется.

Лучше назвать функцию более конкретно и показать в названии, что делает.

File: bot/src/test/java/bot/commandsTest/UntrackCommandTest.java [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+    @Mock
+    private Message message;
+    @Mock
+    private Chat chat;
+    @Mock
+    private Update update;
+    private final UserLinks links = new UserLinks();
+    @InjectMocks
+    private Command untrackCommand = new UntrackCommand(links);
+    private static final User USER = new User(-10L);
```
Comment (iskanred) [inline bot/src/test/java/bot/commandsTest/UntrackCommandTest.java:31]: Магические числа
Лучше подибрать числа норнмальные, либо относящиеся к corner cases (-1, 0, 1, 99 и пр.)

**Comment by iskanred** [comment]
Круто сделал! Нравится твоя архитектура! Комменты добавил, они не критичные и на текущую оценку не влияют. Но хочется видеть исправления в дальнейших домашках либо аргументы, почему исправлять не стоит.

В общем классно постарался, 8/8 💯

File: bot/src/main/java/edu/java/bot/links/UserLinks.java (hunk 5 above)
Comment (onevoker) [reply bot/src/main/java/edu/java/bot/links/UserLinks.java]: Не знал к какому комментарию прикрепить данный вопрос, поэтому прикрепил к этому) Я начал исправлять недочеты и снова вылазит ошибка билда, не понимаю почему, но на гитхабе один тест не проходит, который в классе LinkTest(testCorrectUri), хотя локально тест проходит. Можете подсказать, что с этим делать, пожалуйста.

File: bot/src/main/java/edu/java/bot/links/UserLinks.java (hunk 7 above)
Comment (onevoker) [reply bot/src/main/java/edu/java/bot/links/UserLinks.java]: Может лучше будет поставить аннотацию SneackyThrows, просто в данном методе никогда не будет выброшено исключение, потому что этому классу уже передаются проверенные через Link.isCorrectUri(String link) ссылки, а сам метод просто обрезает лишние слеши в конце ссылки, если вдруг пользователь передаст по сути одинаковые ссылки, но в конце одной из них поставит ,,/,, или например ,,///,,

**Comment by onevoker** [comment]
Я начал исправлять недочеты и у меня снова вылезла ошибка билда, проблема была в классе LinkTest в тесте testCorrectUri. Локально у меня тест проходит, а на гитхабе он падает на ссылках со stackoverflow. Не понимаю, почему так происходит, помогите, пожалуйста)

**Comment by iskanred** [comment]
> Я начал исправлять недочеты и у меня снова вылезла ошибка билда, проблема была в классе LinkTest в тесте testCorrectUri. Локально у меня тест проходит, а на гитхабе он падает на ссылках со stackoverflow. Не понимаю, почему так происходит, помогите, пожалуйста)

после проверки дз могу не увидеть, что ты в гихабе написал, лучше можешь сходить с вопросом в личку телеги

File: bot/src/main/java/edu/java/bot/links/UserLinks.java (hunk 7 above)
Comment (iskanred) [reply bot/src/main/java/edu/java/bot/links/UserLinks.java]: можешь sneakyThrows попробовать

ну как минимум точно стоит хотя бы коммент оставить, почему ты бросаешь `RuntimeException`, а еще лучше указать в документации к методу, потому что, например, я сходу не понял, видишь, пришлось уточнять, тратить время)

File: bot/src/main/java/edu/java/bot/processor/UserMessageProcessorImpl.java [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+        return listOfCommands;
+    }
+
+    @Override
+    public SendMessage process(Update update) {
+        Message message = update.message();
+        String response = message.text();
+
+        for (var command : this.commands()) {
+            if (response != null && response.startsWith(command.command())) {
```
Comment (iskanred) [inline bot/src/main/java/edu/java/bot/processor/UserMessageProcessorImpl.java:41]: Вдруг в дальнейшем появятся команды с одинаковым префексом? Лучше развить проверку

File: bot/src/main/java/edu/java/bot/processor/UserMessageProcessorImpl.java [hunk 16]
```diff
@@ -36,9 +37,11 @@ public List<? extends Command> commands() {
     public SendMessage process(Update update) {
         Message message = update.message();
         String response = message.text();
+        String[] responseArr = response.split(" ");
+        String strCommand = responseArr[0];
```
Comment (iskanred) [inline bot/src/main/java/edu/java/bot/processor/UserMessageProcessorImpl.java:41]: массив заводить в отдельную переменную класть необязательно, можно просто
```
String inputCommand = response.split(" ")[0];
```

Comment (iskanred) [inline bot/src/main/java/edu/java/bot/processor/UserMessageProcessorImpl.java:41]: массив заводить в отдельную переменную класть необязательно, можно просто

```java
String inputCommand = response.split(" ")[0];
```

p.s. имя strCommand не очень информативное

File: bot/src/main/java/edu/java/bot/processor/UserMessageProcessorImpl.java [hunk 17]
```diff
@@ -36,9 +37,11 @@ public List<? extends Command> commands() {
...
 
         for (var command : this.commands()) {
-            if (response != null && response.startsWith(command.command())) {
+            if (Objects.equals(strCommand, command.command())) {
```
Comment (iskanred) [inline bot/src/main/java/edu/java/bot/processor/UserMessageProcessorImpl.java:44]: можно же сравнить просто
```java
s1.equals(s2)
```

**Review by iskanred — CHANGES_REQUESTED** [review]
Оставил комменты, посылаю запрос на доработку

File: bot/src/main/java/edu/java/bot/processor/UserMessageProcessorImpl.java (hunk 17 above)
Comment (onevoker) [reply bot/src/main/java/edu/java/bot/processor/UserMessageProcessorImpl.java]: ✅

**Review by iskanred — APPROVED** [review]
💯
